batch/batch/driver/instance_collection/job_private.py

```diff
@@ -349,54 +350,62 @@ async def create_instances_loop_body(self):
         }
 
         async def user_runnable_jobs(user, remaining) -> AsyncIterator[Dict[str, Any]]:
-            async for batch in self.db.select_and_fetchall(
+            async for job_group in self.db.select_and_fetchall(
                 """
-SELECT batches.id, job_groups_cancelled.id IS NOT NULL AS cancelled, userdata, user, format_version
-FROM batches
-LEFT JOIN job_groups_cancelled
-       ON batches.id = job_groups_cancelled.id
-WHERE user = %s AND `state` = 'running';
+SELECT job_groups.batch_id, job_groups.job_group_id, t.cancelled IS NOT NULL AS cancelled, userdata, job_groups.user, format_version
+FROM job_groups
+LEFT JOIN batches ON batches.id = job_groups.batch_id
+LEFT JOIN LATERAL (
+  SELECT 1 AS cancelled
+  FROM job_group_self_and_ancestors
+  INNER JOIN job_groups_cancelled
+    ON job_group_self_and_ancestors.batch_id = job_groups_cancelled.id AND
+      job_group_self_and_ancestors.ancestor_id = job_groups_cancelled.job_group_id
+  WHERE job_groups.batch_id = job_group_self_and_ancestors.batch_id AND
+    job_groups.job_group_id = job_group_self_and_ancestors.job_group_id
+) AS t ON TRUE
+WHERE job_groups.user = %s AND job_groups.`state` = 'running';
 """,
                 (user,),
             ):
                 async for record in self.db.select_and_fetchall(
                     """
 SELECT jobs.batch_id, jobs.job_id, jobs.spec, jobs.cores_mcpu, regions_bits_rep, COALESCE(SUM(instances.state IS NOT NULL AND
-  (instances.state = 'pending' OR instances.state = 'active')), 0) as live_attempts
+  (instances.state = 'pending' OR instances.state = 'active')), 0) as live_attempts, jobs.job_group_id
 FROM jobs FORCE INDEX(jobs_batch_id_state_always_run_inst_coll_cancelled)
 LEFT JOIN attempts ON jobs.batch_id = attempts.batch_id AND jobs.job_id = attempts.job_id
 LEFT JOIN instances ON attempts.instance_name = instances.name
-WHERE jobs.batch_id = %s AND jobs.state = 'Ready' AND always_run = 1 AND jobs.inst_coll = %s
+WHERE jobs.batch_id = %s AND jobs.job_group_id = %s AND jobs.state = 'Ready' AND always_run = 1 AND jobs.inst_coll = %s
 GROUP BY jobs.job_id, jobs.spec, jobs.cores_mcpu
 HAVING live_attempts = 0
 LIMIT %s;
 """,
-                    (batch['id'], self.name, remaining.value),
+                    (job_group['batch_id'], job_group['job_group_id'], self.name, remaining.value),
                 ):
-                    record['batch_id'] = batch['id']
-                    record['userdata'] = batch['userdata']
-                    record['user'] = batch['user']
-                    record['format_version'] = batch['format_version']
+                    record['batch_id'] = job_group['batch_id']
+                    record['userdata'] = job_group['userdata']
+                    record['user'] = job_group['user']
+                    record['format_version'] = job_group['format_version']
                     yield record
-                if not batch['cancelled']:
+                if not job_group['cancelled']:
```

Comment:
https://github.com/hail-is/hail/pull/14170/files#r1476848548

OK, let's not change in this PR (let's keep it as small as possible!), but let's make an issue to revisit this. This code is wordy enough to make it hard to understand and review. It also appears we may have inadvertently used the wrong index in one case (see below comment).

---

My concrete critique:

The two queries are exactly the same except for the index issue and these two `WHERE` clauses:

```sql
... AND always_run = 1 AND jobs.inst_coll = %s
```
```sql
... AND always_run = 0 AND jobs.inst_coll = %s AND cancelled = 0
```

but it is really hard to see that difference buried among the other nine lines. Moreover, the two
differing where clauses are separated by a non-differing clause. AFAICT, there is no compelling
reason to separately query for always run jobs and then for jobs whose dependencies have not
failed. Why not grab everything that applies in one query?

```python3
async for record in self.db.select_and_fetchall(f"""
SELECT
  jobs.batch_id,
  jobs.job_id,
  jobs.spec,
  jobs.cores_mcpu,
  regions_bits_rep,
  COALESCE(SUM(instances.state IS NOT NULL AND (instances.state = 'pending' OR instances.state = 'active')), 0) as live_attempts,
  jobs.job_group_id
FROM jobs FORCE INDEX(jobs_batch_id_state_always_run_inst_coll_cancelled)
LEFT JOIN attempts ON jobs.batch_id = attempts.batch_id AND jobs.job_id = attempts.job_id
LEFT JOIN instances ON attempts.instance_name = instances.name
WHERE jobs.batch_id = %s
  AND jobs.job_group_id = %s
  AND jobs.inst_coll = %s
  AND jobs.state = 'Ready'
  AND (always_run OR NOT %s)
GROUP BY jobs.job_id, jobs.spec, jobs.cores_mcpu
HAVING live_attempts = 0
LIMIT %s;
""",
    (job_group['batch_id'], job_group['job_group_id'], job_group['cancelled'], remaining.value),
):
    yield record
```

AFAICT, this split started [back when](
https://github.com/hail-is/hail/pull/7933/files#diff-0e49bee3d562ef03a3e157729b60af4314e69840b290e46cc17f23fb7cebc3b5R296-R328)
this whole thing was about thirty lines and the queries were very short and easy to distinguish. It
seems now is as good a time as any to revisit the code and look for an easier to read alternative.

Reply:
Will make an issue and link here when created.